engine/include/cubos/engine/ui/canvas/ui_element.hpp

```diff
@@ -0,0 +1,43 @@
+/// @file
+/// @brief Component @ref cubos::engine::UIElement.
+/// @ingroup canvas-plugin
+
+#pragma once
+
+#include <glm/vec2.hpp>
+
+#include <cubos/core/gl/render_device.hpp>
+#include <cubos/core/reflection/reflect.hpp>
+
+#include <cubos/engine/api.hpp>
+
+using namespace cubos::core::gl;
+
+namespace cubos::engine
+{
+    /// @brief Component which defines the space an element of the UI occupies as a rect.
+    /// @ingroup canvas-plugin
+    struct CUBOS_ENGINE_API UIElement
+    {
+        CUBOS_REFLECT;
+
+        glm::vec2 offset = {0.0F, 0.0F}; ///< Offset of the pivot.
+        glm::vec2 size = {0.0F, 0.0F};   ///< Size of the element.
+        glm::vec2 pivot = {0.5F, 0.5F};  ///< Relative position of the pivot to the element itself.
+        glm::vec2 anchor = {0.5F, 0.5F}; ///< Relative position of the pivot to the parent element.
+        int layer = 0;                   ///< Rendering order layer.
+
+        glm::vec2 position = {0.0F, 0.0F}; ///< Global position of the pivot.
+        glm::mat4 vp;                      ///< View matrix provided by the canvas.
+
+        int hierarchyDepth = 0; ///< Depth in hierarchy. Used for depth sorting.
```

Comment:
I don't like that it isn't clear which fields are supposed to be kept by the user and which are managed by the plugin.
Any ideas on how to make this separation? This is a problem that's emerging in other plugins as well, we should try to have a 'standard solution'.

Reply:
change name to DONT_TOUCH_THIS_ITS_MEANT_TO_BE_INTERNAL_hierarchyDepth 😌